### DISCUSSION TITLE
Tmux binding

### DISCUSSION THREAD

**Discussion by LiamKearn** [fish-shell/fish-shell#9095] [Q&A] [answered]
I should probably just look at someones tmux conf but..

I've got a small function that opens tmux:
```fish
if status is-interactive
    function launchtmux
        if not test -n "$TMUX"
            tmux
        end
    end
    bind \ct launchtmux
end
```
when I exit tmux it requires that I input another character to see my prompt. eg:
<img width="988" alt="Screenshot 2022-07-29" src="https://user-images.githubusercontent.com/76269376/181755243-79070c3d-9ddc-4223-a68a-9da72ba228be.png">
<img width="921" alt="Screenshot 2022-07-29" src="https://user-images.githubusercontent.com/76269376/181755548-4abf724e-9072-4a6a-a441-2025f536365f.png">

Any pointers? Thanks :-)

**Comment by LiamKearn** [comment]
I should note if I run `tmux` myself in fish shell exiting works fine!

**Comment by faho** [comment, marked as answer]
The problem is that running things via bindings goes around some setup that fish will do for interactively run programs - there's no way of knowing if what you are running is something interactive that goes fullscreen like tmux or something like `true` that just does stuff and returns a status.

So you'll have to redo some things you would otherwise get for free.

At the very least, you want `commandline -f repaint`:

```fish
if status is-interactive
    function launchtmux
        if not test -n "$TMUX"
            tmux
            commandline -f repaint
        end
    end
    bind \ct launchtmux
end
```

This will tell fish to redraw the prompt.

But what you *really* want is to just actually *run* `tmux` as if you had typed it in the commandline:

```fish
if status is-interactive
    function launchtmux
        if not test -n "$TMUX"
            commandline -r -- tmux
            commandline -f execute
        end
    end
    bind \ct launchtmux
end
```

(tbh I'd nix the keybinding entirely and make an abbreviation `t` - `abbr --add t 'test -z "$TMUX"; and tmux'`)

**Reply by LiamKearn** [reply]
This is exactly what I'm after. https://fishshell.com/docs/current/cmds/commandline.html is super handy to know about.

I think the only other time I've seen it (without looking into it) is in ports of bash's positional history expansion eg. !:2 which is probably the only thing I miss from bash. Cycling up and down with fuzz is far more user friendly (fish-like) but sometimes if you already know you're after the bash expansion does work well.

Fish has been great so far, Thanks for the work on it :)
